cocotb/drivers/amba.py

```diff
@@ -60,153 +89,403 @@ def __init__(self, entity, name, clock, **kwargs):
+        """Check that the provided burst length is valid."""
+        if length <= 0:
+            raise ValueError("Burst length must be a positive integer")
+
+        if burst is AXIBurst.INCR:
+            if length > 256:
+                raise ValueError("Maximum burst length for INCR bursts is 256")
+        elif burst is AXIBurst.WRAP:
+            if length not in (1, 2, 4, 8, 16):
+                raise ValueError("Burst length for WRAP bursts must be 1, 2, "
+                                 "4, 8 or 16")
+        else:
+            if length > 16:
+                raise ValueError("Maximum burst length for FIXED bursts is 16")
+
+    @staticmethod
+    def _check_size(size: int, data_bus_width: int) -> None:
+        """Check that the provided transfer size is valid."""
+        if size > data_bus_width:
+            raise ValueError("Beat size ({} B) is larger than the bus width "
+                             "({} B)".format(size, data_bus_width))
+        elif size <= 0 or size & (size - 1) != 0:
+            raise ValueError("Beat size must be a positive power of 2")
+
+    @staticmethod
+    def _check_4kB_boundary_crossing(address: int, burst: AXIBurst, size: int,
+                                     length: int) -> None:
+        """Check that the provided burst does not cross a 4kB boundary."""
+        if burst is AXIBurst.INCR:
+            last_address = address + size * (length - 1)
+            if address & ~0xfff != last_address & ~0xfff:
+                raise ValueError(
+                    "INCR burst with start address {:#x} and last address "
+                    "{:#x} crosses the 4kB boundary {:#x}, which is forbidden "
+                    "in a single burst"
+                    .format(address, last_address,
+                            (address & ~0xfff) + 0x1000))
+
+    async def _send_write_address(
+        self, address: int, length: int, burst: AXIBurst, size: int,
+        delay: int, sync: bool
+    ) -> None:
+        """Send the write address, with optional delay (in clocks)"""
+        async with self.write_address_busy:
+            if sync:
+                await RisingEdge(self.clock)
+
+            await ClockCycles(self.clock, delay)
+
+            # Set the address and, if present on the bus, burst, length and
+            # size
+            self.bus.AWADDR <= address
+            self.bus.AWVALID <= 1
+
+            if hasattr(self.bus, "AWBURST"):
+                self.bus.AWBURST <= burst.value
+
+            if hasattr(self.bus, "AWLEN"):
+                self.bus.AWLEN <= length - 1
+
+            if hasattr(self.bus, "AWSIZE"):
+                self.bus.AWSIZE <= size.bit_length() - 1
+
+            # Wait until acknowledged
+            while True:
+                await ReadOnly()
+                if self.bus.AWREADY.value:
+                    break
+                await RisingEdge(self.clock)
             await RisingEdge(self.clock)
+            self.bus.AWVALID <= 0
```

Comment:
```suggestion
            while True:
                await ReadOnly()
                if self.bus.AWREADY.value:
                    break
            await RisingEdge(self.clock)
            self.bus.AWVALID <= 0
```
I don't think you need the `RisingEdge` in the loop. The coroutine may run on more timesteps but you should see the *AWREADY* value at the end of the last timestep before the edge of the clock.

Alternatively, couldn't you just do this:
```py3
            while True:
                yield RisingEdge(self.clock)
                if self.bus.AWREADY.value:
                    break
            self.bus.AWVALID <= 0
```
Either would allow back to back writes assuming you don't use sync.